modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp

```diff
@@ -1,562 +1,601 @@
+    , totalFragmentQuery_{0}
+    , clear_("oit/simplequad.vert", "oit/clear.frag", false)
+    , display_("oit/simplequad.vert", "oit/display.frag", false)
+    , illustration_{screenSize_, fragmentSize_} {
+
+    buildShaders();
+
+    illustrationOnReload_ = illustration_.onReload.add([this]() { onReload_.invoke(); });
+    clear_.onReload([this]() { onReload_.invoke(); });
+    display_.onReload([this]() { onReload_.invoke(); });
+
+    abufferIdxTex_.initialize(nullptr);
+
+    // create fragment query
+    glGenQueries(1, &totalFragmentQuery_);
+    LGL_ERROR;
+}
+
+FragmentListRenderer::~FragmentListRenderer() {
+    if (totalFragmentQuery_) glDeleteQueries(1, &totalFragmentQuery_);
+}
+
+void FragmentListRenderer::prePass(const size2_t& screenSize) {
+    resizeBuffers(screenSize);
+
+    // reset counter
+
+    GLuint v[1] = {0};
+    atomicCounter_.upload(v, sizeof(GLuint));
+    atomicCounter_.unbind();
+
+    // clear textures
+    clear_.activate();
+    auto& texUnit = textureUnits_.emplace_back();
+    setUniforms(clear_, texUnit);
+
+    utilgl::GlBoolState depthTest(GL_DEPTH_TEST, GL_TRUE);
+    utilgl::DepthMaskState depthMask(GL_TRUE);
+    utilgl::DepthFuncState depthFunc(GL_ALWAYS);
+    utilgl::CullFaceState culling(GL_NONE);
+    utilgl::singleDrawImagePlaneRect();
+
+    clear_.deactivate();
+
+    // memory barrier
+    glMemoryBarrier(GL_TEXTURE_FETCH_BARRIER_BIT | GL_ATOMIC_COUNTER_BARRIER_BIT);
+
+    // start query
+    // The query is used to determinate the size needed for the shader storage buffer
+    // to store all the fragments.
+    glBeginQuery(GL_SAMPLES_PASSED, totalFragmentQuery_);
+    LGL_ERROR;
+}
+
+bool FragmentListRenderer::postPass(bool useIllustration) {
+    // memory barrier
+    glMemoryBarrier(GL_TEXTURE_FETCH_BARRIER_BIT | GL_SHADER_STORAGE_BARRIER_BIT);
+
+    // get query result
+    GLuint numFrags = 0;
+    glEndQuery(GL_SAMPLES_PASSED);
+    glGetQueryObjectuiv(totalFragmentQuery_, GL_QUERY_RESULT, &numFrags);
+    LGL_ERROR;
+
+    // check if enough space was available
+    if (numFrags > fragmentSize_) {
+        // we have to resize the fragment storage buffer
+        LogInfo("fragment lists resolved, pixels drawn: "
+                << numFrags << ", available: " << fragmentSize_ << ", allocate space for "
+                << int(1.1f * numFrags) << " pixels");
+        fragmentSize_ = static_cast<size_t>(1.1f * numFrags);
+
+        // unbind texture
+        textureUnits_.clear();
+        return false;
+    }
+
+    // Set depth buffer to read from.
+    auto bg = backgroundPort_.lock();
+    bool hasBackground = bg && bg->hasData();
+
+    // Build shader depending on inport state.
+    if (hasBackground != builtWithBackground_) buildShaders();
+
+    if (!useIllustration) {
+        // render fragment list
+        display_.activate();
+        setUniforms(display_, textureUnits_[0]);
+        if (builtWithBackground_) {

```

Comment:
it is just rebuilt above if background exists. Is it possible that the building does not include background anyway? In that case it will be rebuilt every frame?

Reply:
@AnkeAnke ping